### DISCUSSION TITLE
Adding Macro Menu items from external file

### DISCUSSION THREAD

**Discussion by Regretfully** [eteran/nedit-ng#319] [Q&A]
Hi!

Is it possible to add macro **menu items** using the autoload.nm file? defining subroutines work fine but when i try to create a macro menu item i get syntax error. I am trying to achieve a way to roll out tools to multiple users without having to mess with their nedit.rc file to add macro menu items, appended sample code below:

## autoload.nm ###
load_macro_file("/home/a394186/.nedit/macrotest.rc")
###############

## macrotest.rc ####
nedit.macroCommands: \
	Test:::: {\n\
		dialog("Testing")\n\
	}\n
################

Kind Regards,

**Comment by eteran** [comment]
Right, I don't think you can use the `autoload.nm` like that but one thing to keep in mind is that NG uses a new multi-file configuration file system found in `$HOME/.config/nedit-ng/`

In there, is a `macros.yaml` so if you want all your uses to have the SAME set of macros, you could just distribute that file alone maybe?

**Reply by Regretfully** [reply]
I want users to be able to keep their own macros (if they have any) and deliver the tool macros to all group users at the same time, maybe an impossible task?

**Reply by eteran** [reply]
I see what you mean. So I don't think we have a separate "user" and "system" macro concept... but maybe we could. I'll explore the best ways to approach this.

**Reply by eteran** [reply]
So my thoughts on this is that we could have an environment variable like: `NEDIT_GLOBAL_CONFIG` that would point to a directory that has the base "system" settings. And that the user's settings would be basically overlayed on top of that.

The main tricky parts will be when the user changes settings. We'll need a good way to untangle system settings from user settings so we only write the ones we need to. Some of this is easy enough, but there are some interesting corner cases. Like if the user loads the system macros, and then **deletes** one that was in the system macro set.

I can imagine just saying "OK, the user now has a `macros.yaml` that has all of what was in the system, but has their local changes applied. That'll work great... but now if you change the system ones they won't see it automatically! So... it gets complicated. Maybe we just don't allow users to edit settings that were specified in the system ones? I'm honestly not sure.

Either way, I think this would be a very useful feature, and frankly you're not the first user to touch on this subject. So we'll see what we can do.

**Reply by Regretfully** [reply]
I am not a nedit-ng user so maybe I am missing something but wouldnt it be possible to load the system macros in the same way you can use autoload.nm to load subroutines every time nedit starts?

I have been confused and agitated by the fact that you cannot sideload macros in nedit for a couple of weeks. Unfortunately the documentation is very lackluster in nedit and the user-base is (understandably) small.

**Reply by eteran** [reply]
I can certainly look into some sort of side loading system of that kind to see if it is doable without major changes.

**Reply by Regretfully** [reply]
For posterity's sake:
launching nedit with "nedit -import filewithmacros" does pretty much exactly what I want. Except it would be nice to be able to have it in the autoload.nm file somehow.

**Reply by eteran** [reply]
Right, and that would work in nedit-ng too I believe. However, my only concern with that approach is that the moment a user saves their settings, they will save the combination of their settings and the macros back into their settings.

This may be no problem at all, but it does mean that if you update `filewithmacros` in the future, they may not see them, or possibly they'd see both.

This is because it's not really designed as a feature to include settings so much as it is to import settings.

**Reply by eteran** [reply]
For the record, I do like your `load_macro_file` idea though and will likely implement it :-)

**Reply by Regretfully** [reply]
I agree about the problems with saving their settings... I might have to find some other workaround, perhaps a wrapper that parses the users nedit.rc and checks for updates to the macro items... Thank you very much for the discussion. Its not easy these days to reach any community working with nedit.
